File: src/harness_core/artifact_gate.py
"""Artifact Gate for Stage 2 — verify task artifacts deterministically."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .task_records import TaskRecordBundle
from .validators import validate_completion_record, validate_handoff_pack


@dataclass(frozen=True)
class ArtifactCheck:
    name: str
    passed: bool
    message: str


@dataclass(frozen=True)
class ArtifactGateResult:
    ok: bool
    checks: tuple[ArtifactCheck, ...]
    missing_artifacts: tuple[str, ...]
    schema_violations: tuple[str, ...]
    forbidden_violations: tuple[str, ...]


class ArtifactGate:
    """Verify artifact existence, schema, evidence refs, and file policy."""

    def evaluate(
        self,
        bundle: TaskRecordBundle,
        allowed_files: tuple[str, ...] | None = None,
        forbidden_files: tuple[str, ...] | None = None,
    ) -> ArtifactGateResult:
        checks: list[ArtifactCheck] = []
        missing: list[str] = []
        schema_violations: list[str] = []
        forbidden: list[str] = []

        self._check_completion_schema(bundle, checks, schema_violations)
        self._check_handoff_schema(bundle, checks, schema_violations)
        self._check_artifact_existence(bundle, checks, missing)
        self._check_evidence_refs(bundle, checks, missing)
        self._check_handoff_pack_ref(bundle, checks, missing)
        self._check_allowed_files(bundle, allowed_files, checks, missing)
        self._check_forbidden_files(bundle, forbidden_files, checks, forbidden)

        ok = all(c.passed for c in checks)
        return ArtifactGateResult(
            ok=ok,
            checks=tuple(checks),
            missing_artifacts=tuple(missing),
            schema_violations=tuple(schema_violations),
            forbidden_violations=tuple(forbidden),
        )

    def _check_completion_schema(
        self,
        bundle: TaskRecordBundle,
        checks: list[ArtifactCheck],
        violations: list[str],
    ) -> None:
        result = validate_completion_record(bundle.completion)
        if result.ok:
            checks.append(ArtifactCheck("completion_schema", True, "completion.json valid"))
        else:
            msg = "; ".join(result.errors)
            violations.append(msg)
            checks.append(ArtifactCheck("completion_schema", False, msg))

    def _check_handoff_schema(
        self,
        bundle: TaskRecordBundle,
        checks: list[ArtifactCheck],
        violations: list[str],
    ) -> None:
        result = validate_handoff_pack(bundle.handoff_pack)
        if result.ok:
            checks.append(ArtifactCheck("handoff_schema", True, "handoff_pack.json valid"))
        else:
            msg = "; ".join(result.errors)
            violations.append(msg)
            checks.append(ArtifactCheck("handoff_schema", False, msg))

    def _check_artifact_existence(
        self,
        bundle: TaskRecordBundle,
        checks: list[ArtifactCheck],
        missing: list[str],
    ) -> None:
        refs = bundle.completion.get("artifact_refs", [])
        if not isinstance(refs, list) or not refs:
            checks.append(ArtifactCheck("artifact_existence", True, "no artifact_refs to check"))
            return

        for ref in refs:
            path = ref.get("path", "") if isinstance(ref, dict) else ""
            if not path:
                missing.append("<empty path>")
                checks.append(ArtifactCheck("artifact_existence", False, "artifact_ref has no path"))
                continue
            full_path = bundle.task_dir / path
            if not full_path.exists():
                missing.append(path)
                checks.append(ArtifactCheck("artifact_existence", False, f"not found: {path}"))

        if all(c.passed for c in checks if c.name == "artifact_existence"):
            checks.append(ArtifactCheck("artifact_existence", True, "all artifacts exist"))

    def _check_evidence_refs(
        self,
        bundle: TaskRecordBundle,
        checks: list[ArtifactCheck],
        missing: list[str],
    ) -> None:
        refs = bundle.handoff_pack.get("evidence_refs", [])
        if not isinstance(refs, list) or not refs:
            checks.append(ArtifactCheck("evidence_refs", False, "evidence_refs empty or missing"))
            missing.append("evidence_refs")
            return

        all_ok = True
        for ref in refs:
            if not isinstance(ref, dict) or not ref.get("path"):
                all_ok = False
                missing.append("<invalid evidence_ref>")

        if all_ok:
            checks.append(ArtifactCheck("evidence_refs", True, "evidence_refs valid"))
        else:
            checks.append(ArtifactCheck("evidence_refs", False, "some evidence_refs invalid"))

    def _check_handoff_pack_ref(
        self,
        bundle: TaskRecordBundle,
        checks: list[ArtifactCheck],
        missing: list[str],
    ) -> None:
        ref = bundle.completion.get("handoff_pack_ref")
        if not ref:
            checks.append(ArtifactCheck("handoff_pack_ref", True, "no handoff_pack_ref to check"))
            return
        full_path = bundle.task_dir / ref
        if full_path.exists():
            checks.append(ArtifactCheck("handoff_pack_ref", True, f"handoff_pack_ref exists: {ref}"))
        else:
            missing.append(ref)
            checks.append(ArtifactCheck("handoff_pack_ref", False, f"handoff_pack_ref not found: {ref}"))

    def _check_allowed_files(
        self,
        bundle: TaskRecordBundle,
        allowed_files: tuple[str, ...] | None,
        checks: list[ArtifactCheck],
        missing: list[str],
    ) -> None:
        if allowed_files is None:
            checks.append(ArtifactCheck("allowed_files", True, "no allowed_files constraint"))
            return

        artifact_paths = []
        for ref in bundle.completion.get("artifact_refs", []):
            if isinstance(ref, dict) and ref.get("path"):
                artifact_paths.append(ref["path"])

        uncovered = [p for p in artifact_paths if not any(p.startswith(a) or a.startswith(p) for a in allowed_files)]
        if not uncovered:
            checks.append(ArtifactCheck("allowed_files", True, "all artifacts covered by allowed_files"))
        else:
            for p in uncovered:
                missing.append(f"not in allowed_files: {p}")
            checks.append(ArtifactCheck("allowed_files", False, f"{len(uncovered)} artifact(s) not in allowed_files"))

    def _check_forbidden_files(
        self,
        bundle: TaskRecordBundle,
        forbidden_files: tuple[str, ...] | None,
        checks: list[ArtifactCheck],
        forbidden: list[str],
    ) -> None:
        if forbidden_files is None:
            checks.append(ArtifactCheck("forbidden_files", True, "no forbidden_files constraint"))
            return

        violations = []
        for ref in bundle.completion.get("artifact_refs", []):
            if isinstance(ref, dict) and ref.get("path"):
                path = ref["path"]
                if any(path.startswith(f) or f.startswith(path) for f in forbidden_files):
                    violations.append(path)

        if not violations:
            checks.append(ArtifactCheck("forbidden_files", True, "no forbidden_files violations"))
        else:
            forbidden.extend(violations)
            checks.append(ArtifactCheck("forbidden_files", False, f"{len(violations)} forbidden violation(s)"))

File: engine/src/quality/artifact_gate.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

// ---------------------------------------------------------------------------
// ArtifactCheck
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ArtifactCheck {
    pub name: String,
    pub passed: bool,
    pub message: String,
}

impl Default for ArtifactCheck {
    fn default() -> Self {
        Self {
            name: String::new(),
            passed: false,
            message: String::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// ArtifactGateResult
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ArtifactGateResult {
    pub ok: bool,
    pub checks: Vec<ArtifactCheck>,
    pub missing_artifacts: Vec<String>,
    pub schema_violations: Vec<String>,
    pub forbidden_violations: Vec<String>,
}

impl Default for ArtifactGateResult {
    fn default() -> Self {
        Self {
            ok: false,
            checks: Vec::new(),
            missing_artifacts: Vec::new(),
            schema_violations: Vec::new(),
            forbidden_violations: Vec::new(),
        }
    }
}

// ---------------------------------------------------------------------------
// ArtifactGate
// ---------------------------------------------------------------------------

pub struct ArtifactGate;

impl Default for ArtifactGate {
    fn default() -> Self {
        Self
    }
}

impl ArtifactGate {
    pub fn new() -> Self {
        Self
    }

    pub fn evaluate(
        &self,
        completion: &Value,
        handoff_pack: &Value,
        allowed_files: Option<&[String]>,
        forbidden_files: Option<&[String]>,
    ) -> ArtifactGateResult {
        let mut checks = Vec::new();
        let mut missing = Vec::new();
        let mut schema_violations = Vec::new();
        let mut forbidden = Vec::new();

        self.check_completion_schema(completion, &mut checks, &mut schema_violations);
        self.check_handoff_schema(handoff_pack, &mut checks, &mut schema_violations);
        self.check_artifact_existence(completion, &mut checks, &mut missing);
        self.check_evidence_refs(handoff_pack, &mut checks, &mut missing);
        self.check_handoff_pack_ref(completion, &mut checks, &mut missing);
        self.check_allowed_files(completion, allowed_files, &mut checks, &mut missing);
        self.check_forbidden_files(completion, forbidden_files, &mut checks, &mut forbidden);

        let ok = checks.iter().all(|c| c.passed);

        ArtifactGateResult {
            ok,
            checks,
            missing_artifacts: missing,
            schema_violations,
            forbidden_violations: forbidden,
        }
    }

    fn check_completion_schema(
        &self,
        completion: &Value,
        checks: &mut Vec<ArtifactCheck>,
        violations: &mut Vec<String>,
    ) {
        if !completion.is_object() {
            let msg = "completion must be a JSON object".to_string();
            violations.push(msg.clone());
            checks.push(ArtifactCheck {
                name: "completion_schema".to_string(),
                passed: false,
                message: msg,
            });
            return;
        }

        let obj = completion.as_object().unwrap();
        if obj.get("status").and_then(Value::as_str).is_none() {
            let msg = "completion missing 'status' field".to_string();
            violations.push(msg.clone());
            checks.push(ArtifactCheck {
                name: "completion_schema".to_string(),
                passed: false,
                message: msg,
            });
            return;
        }

        checks.push(ArtifactCheck {
            name: "completion_schema".to_string(),
            passed: true,
            message: "completion valid".to_string(),
        });
    }

    fn check_handoff_schema(
        &self,
        handoff_pack: &Value,
        checks: &mut Vec<ArtifactCheck>,
        violations: &mut Vec<String>,
    ) {
        if !handoff_pack.is_object() {
            let msg = "handoff_pack must be a JSON object".to_string();
            violations.push(msg.clone());
            checks.push(ArtifactCheck {
                name: "handoff_schema".to_string(),
                passed: false,
                message: msg,
            });
            return;
        }

        checks.push(ArtifactCheck {
            name: "handoff_schema".to_string(),
            passed: true,
            message: "handoff_pack valid".to_string(),
        });
    }

    fn check_artifact_existence(
        &self,
        completion: &Value,
        checks: &mut Vec<ArtifactCheck>,
        missing: &mut Vec<String>,
    ) {
        let refs = match completion.get("artifact_refs").and_then(Value::as_array) {
            Some(r) if !r.is_empty() => r,
            _ => {
                checks.push(ArtifactCheck {
                    name: "artifact_existence".to_string(),
                    passed: true,
                    message: "no artifact_refs to check".to_string(),
                });
                return;
            }
        };

        let mut any_failed = false;
        for ref_val in refs {
            let path = ref_val
                .as_object()
                .and_then(|o| o.get("path"))
                .and_then(Value::as_str)
                .unwrap_or("");

            if path.is_empty() {
                missing.push("<empty path>".to_string());
                checks.push(ArtifactCheck {
                    name: "artifact_existence".to_string(),
                    passed: false,
                    message: "artifact_ref has no path".to_string(),
                });
                any_failed = true;
                continue;
            }

            if !std::path::Path::new(path).exists() {
                missing.push(path.to_string());
                checks.push(ArtifactCheck {
                    name: "artifact_existence".to_string(),
                    passed: false,
                    message: format!("not found: {}", path),
                });
                any_failed = true;
            }
        }

        if !any_failed {
            checks.push(ArtifactCheck {
                name: "artifact_existence".to_string(),
                passed: true,
                message: "all artifacts exist".to_string(),
            });
        }
    }

    fn check_evidence_refs(
        &self,
        handoff_pack: &Value,
        checks: &mut Vec<ArtifactCheck>,
        missing: &mut Vec<String>,
    ) {
        let refs = match handoff_pack.get("evidence_refs").and_then(Value::as_array) {
            Some(r) if !r.is_empty() => r,
            _ => {
                checks.push(ArtifactCheck {
                    name: "evidence_refs".to_string(),
                    passed: false,
                    message: "evidence_refs empty or missing".to_string(),
                });
                missing.push("evidence_refs".to_string());
                return;
            }
        };

        let all_ok = refs.iter().all(|ref_val| {
            ref_val
                .as_object()
                .and_then(|o| o.get("path"))
                .and_then(Value::as_str)
                .map(|s| !s.is_empty())
                .unwrap_or(false)
        });

        if all_ok {
            checks.push(ArtifactCheck {
                name: "evidence_refs".to_string(),
                passed: true,
                message: "evidence_refs valid".to_string(),
            });
        } else {
            missing.push("<invalid evidence_ref>".to_string());
            checks.push(ArtifactCheck {
                name: "evidence_refs".to_string(),
                passed: false,
                message: "some evidence_refs invalid".to_string(),
            });
        }
    }

    fn check_handoff_pack_ref(
        &self,
        completion: &Value,
        checks: &mut Vec<ArtifactCheck>,
        missing: &mut Vec<String>,
    ) {
        let ref_path = match completion.get("handoff_pack_ref").and_then(Value::as_str) {
            Some(r) if !r.is_empty() => r,
            _ => {
                checks.push(ArtifactCheck {
                    name: "handoff_pack_ref".to_string(),
                    passed: true,
                    message: "no handoff_pack_ref to check".to_string(),
                });
                return;
            }
        };

        if std::path::Path::new(ref_path).exists() {
            checks.push(ArtifactCheck {
                name: "handoff_pack_ref".to_string(),
                passed: true,
                message: format!("handoff_pack_ref exists: {}", ref_path),
            });
        } else {
            missing.push(ref_path.to_string());
            checks.push(ArtifactCheck {
                name: "handoff_pack_ref".to_string(),
                passed: false,
                message: format!("handoff_pack_ref not found: {}", ref_path),
            });
        }
    }

    fn check_allowed_files(
        &self,
        completion: &Value,
        allowed_files: Option<&[String]>,
        checks: &mut Vec<ArtifactCheck>,
        missing: &mut Vec<String>,
    ) {
        let allowed = match allowed_files {
            Some(a) => a,
            None => {
                checks.push(ArtifactCheck {
                    name: "allowed_files".to_string(),
                    passed: true,
                    message: "no allowed_files constraint".to_string(),
                });
                return;
            }
        };

        let artifact_paths: Vec<String> = completion
            .get("artifact_refs")
            .and_then(Value::as_array)
            .map(|refs| {
                refs.iter()
                    .filter_map(|r| {
                        r.as_object()
                            .and_then(|o| o.get("path"))
                            .and_then(Value::as_str)
                            .map(String::from)
                    })
                    .collect()
            })
            .unwrap_or_default();

        let uncovered: Vec<&String> = artifact_paths
            .iter()
            .filter(|p| {
                !allowed
                    .iter()
                    .any(|a| p.starts_with(a) || a.starts_with(p.as_str()))
            })
            .collect();

        if uncovered.is_empty() {
            checks.push(ArtifactCheck {
                name: "allowed_files".to_string(),
                passed: true,
                message: "all artifacts covered by allowed_files".to_string(),
            });
        } else {
            for p in &uncovered {
                missing.push(format!("not in allowed_files: {}", p));
            }
            checks.push(ArtifactCheck {
                name: "allowed_files".to_string(),
                passed: false,
                message: format!("{} artifact(s) not in allowed_files", uncovered.len()),
            });
        }
    }

    fn check_forbidden_files(
        &self,
        completion: &Value,
        forbidden_files: Option<&[String]>,
        checks: &mut Vec<ArtifactCheck>,
        forbidden: &mut Vec<String>,
    ) {
        let forbidden_list = match forbidden_files {
            Some(f) => f,
            None => {
                checks.push(ArtifactCheck {
                    name: "forbidden_files".to_string(),
                    passed: true,
                    message: "no forbidden_files constraint".to_string(),
                });
                return;
            }
        };

        let mut violations = Vec::new();
        if let Some(refs) = completion.get("artifact_refs").and_then(Value::as_array) {
            for ref_val in refs {
                if let Some(path) = ref_val
                    .as_object()
                    .and_then(|o| o.get("path"))
                    .and_then(Value::as_str)
                {
                    if forbidden_list
                        .iter()
                        .any(|f| path.starts_with(f) || f.starts_with(path))
                    {
                        violations.push(path.to_string());
                    }
                }
            }
        }

        if violations.is_empty() {
            checks.push(ArtifactCheck {
                name: "forbidden_files".to_string(),
                passed: true,
                message: "no forbidden_files violations".to_string(),
            });
        } else {
            let count = violations.len();
            forbidden.extend(violations);
            checks.push(ArtifactCheck {
                name: "forbidden_files".to_string(),
                passed: false,
                message: format!("{} forbidden violation(s)", count),
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_valid_completion_and_handoff() {
        let gate = ArtifactGate::new();
        let completion = json!({"status": "completed", "exit_code": 0});
        let handoff = json!({"evidence_refs": [{"path": "/tmp/test"}]});

        let result = gate.evaluate(&completion, &handoff, None, None);
        let completion_check = result
            .checks
            .iter()
            .find(|c| c.name == "completion_schema")
            .unwrap();
        assert!(completion_check.passed);
        let handoff_check = result
            .checks
            .iter()
            .find(|c| c.name == "handoff_schema")
            .unwrap();
        assert!(handoff_check.passed);
    }

    #[test]
    fn test_invalid_completion_missing_status() {
        let gate = ArtifactGate::new();
        let completion = json!({"exit_code": 0});
        let handoff = json!({});

        let result = gate.evaluate(&completion, &handoff, None, None);
        assert!(!result.ok);
        assert!(!result.schema_violations.is_empty());
    }

    #[test]
    fn test_missing_evidence_refs() {
        let gate = ArtifactGate::new();
        let completion = json!({"status": "completed"});
        let handoff = json!({});

        let result = gate.evaluate(&completion, &handoff, None, None);
        assert!(result
            .checks
            .iter()
            .any(|c| c.name == "evidence_refs" && !c.passed));
        assert!(result
            .missing_artifacts
            .contains(&"evidence_refs".to_string()));
    }

    #[test]
    fn test_forbidden_file_violation() {
        let gate = ArtifactGate::new();
        let completion = json!({
            "status": "completed",
            "artifact_refs": [{"path": "/etc/passwd"}]
        });
        let handoff = json!({"evidence_refs": [{"path": "/tmp/e"}]});

        let forbidden = vec!["/etc".to_string()];
        let result = gate.evaluate(&completion, &handoff, None, Some(&forbidden));
        assert!(!result.ok);
        assert!(!result.forbidden_violations.is_empty());
    }

    #[test]
    fn test_no_constraints_all_pass() {
        let gate = ArtifactGate::new();
        let completion = json!({"status": "completed"});
        let handoff = json!({"evidence_refs": [{"path": "/tmp/x"}]});

        let result = gate.evaluate(&completion, &handoff, None, None);
        let evidence = result
            .checks
            .iter()
            .find(|c| c.name == "evidence_refs")
            .unwrap();
        assert!(evidence.passed);
    }

    #[test]
    fn test_non_object_handoff_fails() {
        let gate = ArtifactGate::new();
        let completion = json!({"status": "completed"});
        let handoff = json!("not an object");

        let result = gate.evaluate(&completion, &handoff, None, None);
        assert!(!result.ok);
        assert!(result
            .schema_violations
            .iter()
            .any(|v| v.contains("handoff_pack")));
    }

    #[test]
    fn test_result_serializes_roundtrip() {
        let result = ArtifactGateResult {
            ok: true,
            checks: vec![ArtifactCheck {
                name: "test".to_string(),
                passed: true,
                message: "ok".to_string(),
            }],
            missing_artifacts: vec![],
            schema_violations: vec![],
            forbidden_violations: vec![],
        };
        let json_str = serde_json::to_string(&result).unwrap();
        let back: ArtifactGateResult = serde_json::from_str(&json_str).unwrap();
        assert_eq!(result, back);
    }
}
